File: myipnetstat.py
#! /usr/bin/python3

import socket
import sys

class myipud:
    def __init__(self, location):
        self.buffer = ''

        self.s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)

        try:
            self.s.connect(location)

        except FileNotFoundError as e:
            print(f'Socket cannot be opened ({e}), is MyIP running?')
            sys.exit(0)

    def request(self, req):
        req_b = (req + '\n').encode('ascii')

        self.s.send(req_b)

        while True:
            lf = self.buffer.find('\n')

            if lf != -1:
                rc = self.buffer[0:lf]

                self.buffer = self.buffer[lf + 1:]

                return rc

            data = self.s.recv(4096)

            if len(data) == 0:
                rc = self.buffer
                self.buffer = ''
                return rc

            self.buffer += data.decode('ascii')

def help():
    print(' * sessions')
    print(' * list-devices  list the physical devices, to be used with start/stop-pcap')
    print(' * start-pcap x  start pcap recording on device x')
    print(' * stop-pcap x   stop pcap recording on device x')

if len(sys.argv) == 1:
    help()
    sys.exit(1)

mi = myipud('/tmp/myipstats.sock')

if sys.argv[1] == 'sessions':
    print(mi.request('sessions'))

elif sys.argv[1] == 'list-devices':
    print(mi.request('list-devices'))

elif sys.argv[1] == 'start-pcap':
    print(mi.request('start-pcap|%s' % sys.argv[2]))

elif sys.argv[1] == 'stop-pcap':
    print(mi.request('stop-pcap|%s' % sys.argv[2]))

else:
    help();

File: myipnetstat.cpp
#include <jansson.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/un.h>

#include "utils.h"


int main(int argc, char *argv[])
{
	int fd = socket(AF_UNIX, SOCK_STREAM, 0);
	if (fd == -1)
		error_exit(true, "Cannot create unix domain socket");

	struct sockaddr_un remote { 0 };

	remote.sun_family = AF_UNIX;
	strcpy(remote.sun_path, "/tmp/myipstats.sock");

	int len = strlen(remote.sun_path) + sizeof(remote.sun_family);

	if (connect(fd, reinterpret_cast<sockaddr *>(&remote), len) == -1)
		error_exit(true, "Failed to connect");

	char cmd[] = "sessions\n";

	WRITE(fd, reinterpret_cast<const uint8_t *>(cmd), sizeof(cmd) - 1);

	for(;;) {
		char buffer[4096] { 0 };

		int rc = read(fd, buffer, (sizeof buffer) - 1);

		if (rc == 0 || rc == -1)
			break;

		printf("%s\n", buffer);
	}

	close(fd);

	return 0;
}
